### ISSUE TITLE
BackgroundJob.Delete may hang forever

### ISSUE THREAD

**Issue by jods4** [HangfireIO/Hangfire#500]
If you pass a non-existing id to the public API `BackgroundJob.Delete`, it will loop forever.

The bug is in `BackgroundJobStateChanger.GetJobData()` look here:
https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/States/BackgroundJobStateChanger.cs#L130

This method will never stop trying to find a job even if it doesn't exist.

If you look at this comment in caller code, this is clearly not the intended behaviour:
https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/States/BackgroundJobStateChanger.cs#L59-L63

**Comment by gandarez** [comment]
I suppose there's a loop to wait until the the data gets released (unlocked). But I think it might try only twice instead of an infinite loop!

**Comment by jods4** [comment]
@gandarez  I'm not even sure this is the case!

If you look at what `SqlServerConnection.GetJobData()` does under the hood, it comes down to [simplified] this:
```csharp
var jobData = connection.Query<SqlJob>("select * from Job where id = @id", new { id = id }).SingleOrDefault(); 
if (jobData == null) return null; 
```

There is no locking or whatever that justifies waiting here. We plainly look if the job is in the table or not.

Now, I don't fully grasp the Hangfire codebase, maybe those considerations apply to other providers?

**Comment by captain828** [comment]
This is an actual issue I am having right now with the latest Hangfire version (1.5.3) after we migrated our Hangfire database.

It just hangs while attempting to retrieve a job that does not exist and the culprit does seem to be the BackgroundJobStateChanger.GetJobData() method.

@odinserj - please advise.

**Comment by danoleary** [comment]
I ran into this issue yesterday, in 1.5.3. Could BackgroundJobStateChanger.GetJobData() just return null if no job is found for a given id?

**Comment by captain828** [comment]
https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/States/BackgroundJobStateChanger.cs#L59-L63

So the null case should be covered already.

IMHO the https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/BackgroundJobServerOptions.cs#L76 setting or a new setting should be used to set the time-out for getting job data.

Thoughts?

**Comment by danoleary** [comment]
I don't think that does actually cover it, as GetJobData(StateChangeContext context) will get stuck in an infinite loop if context.Connection.GetJobData(context.BackgroundJobId) returns null. 
Adding 
````
    if (jobData == null)
    {
        return null;
    }
````
to GetJobData(StateChangeContext context) should fix the problem though.

**Comment by captain828** [comment]
Yes, I meant to say that a possible null is already covered in `public IState ChangeState(StateChangeContext context)`
It definitely isn't covered in `GetJobData(StateChangeContext context)` due to the infinite loop.

I'll see if I have time later today to do a pull request on this.

**Comment by danoleary** [comment]
I've got time now to create a pull request for this, if you don't have any objections?

**Comment by captain828** [comment]
Sure, go ahead and then reference the pull request here.

Since this is a bug fix, I believe you should be patching on the `master` branch.

**Comment by danoleary** [comment]
Thanks, pull request is here: https://github.com/HangfireIO/Hangfire/pull/535

**Comment by odinserj** [comment]
This issue is already resolved in PR #469. I've just released [1.5.4](https://github.com/HangfireIO/Hangfire/releases/tag/v1.5.4) that fixes the problem.
